flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/util/EqualityAndHashcodeProvider.java

```diff
@@ -0,0 +1,86 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.flink.table.runtime.util;
+
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.functions.FunctionContext;
+import org.apache.flink.table.functions.SpecializedFunction;
+import org.apache.flink.table.types.DataType;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.Serializable;
+import java.lang.invoke.MethodHandle;
+
+import static org.apache.flink.table.api.Expressions.$;
+import static org.apache.flink.table.expressions.ApiExpressionUtils.unresolvedCall;
+
+/** This class is used for scalar function. */
+public class EqualityAndHashcodeProvider implements Closeable, Serializable {
+    private final SpecializedFunction.ExpressionEvaluator hashcodeEvaluator;
+    private final SpecializedFunction.ExpressionEvaluator equalityEvaluator;
+    private static MethodHandle hashcodeHandle;
+
+    private static MethodHandle equalityHandle;
+
+    public EqualityAndHashcodeProvider(
+            SpecializedFunction.SpecializedContext context, DataType dataType) {
+        hashcodeEvaluator =
+                context.createEvaluator(
+                        unresolvedCall(BuiltInFunctionDefinitions.INTERNAL_HASHCODE, $("element1")),
+                        DataTypes.INT(),
+                        DataTypes.FIELD("element1", dataType.notNull().toInternal()));
+
+        equalityEvaluator =
+                context.createEvaluator(
+                        $("element1").isEqual($("element2")),
+                        DataTypes.BOOLEAN(),
+                        DataTypes.FIELD("element1", dataType.notNull().toInternal()),
+                        DataTypes.FIELD("element2", dataType.notNull().toInternal()));
+    }
+
+    public void open(FunctionContext context) throws Exception {
+        hashcodeHandle = hashcodeEvaluator.open(context);
+        equalityHandle = equalityEvaluator.open(context);
+    }
+
+    public static boolean equals(Object o1, Object o2) {
+        try {
+            return (Boolean) equalityHandle.invoke(o1, o2);
+        } catch (Throwable e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static int hashCode(Object o) {
+        try {
+            return (int) hashcodeHandle.invoke(o);
+        } catch (Throwable e) {
+            throw new RuntimeException(e);
+        }
+    }
```

Comment:
This must not be static. You pass a `dataType` in the ctor, which is not a static ctx. Both `equals` and `hashCode` depend on the non-static `dataType`. 

Having `equals` and `hashCode` is asking for troubles.

Reply:
If we don't use static, it will report this error
```
Caused by: java.lang.IncompatibleClassChangeError: Expected static method org.apache.flink.table.runtime.util.EqualityAndHashcodeProvider.hashCode(Ljava/lang/Object;)I
	at org.apache.flink.table.runtime.util.ObjectContainer.hashCode(ObjectContainer.java:45)
	at java.util.HashMap.hash(HashMap.java:340)
	at java.util.HashMap.merge(HashMap.java:1229)
	at org.apache.flink.table.runtime.functions.scalar.ArrayExceptFunction.eval(ArrayExceptFunction.java:75)
	... 11 more
```

Reply:
Sounds correct.

 I'd not couple the `ObjectContainer` with `EqualityAndHashcodeProvider` implementation though. Please extract a separate interface in `ObjectContainer` that `EqualityAndHashcodeProvider` can implement. 
 
`EqualityAndHashcodeProvider` is responsible for initialising the generated code. `ObjectContainer` does not need to know it works with code generation. It just needs implementations of `equals` and `hashcode`. You have two ways to do that:

1.
```
public ObjectContainer(Object o, BiFunction<Object, Object, Integer> equalsMethod, Function<Object, Integer> hashCodeMethod) {
 ....
}

new ObjectContainer(o, equalsAndHashCodeProvider::equals, equalsAndHashCodeProvider::hashcode);
```

2. 
```
class ObjectContainer {
interface EqualsAndHashcode {
  int equals(Object o1, Object o2);
  int hashcode(Object o);
}

public ObjectContainer(Object o, EqualsAndHashcode equalsAndHashcode) 
...{
}

EqualsAndHashcodeProvider implements EqualsAndHashcode
```

Reply:
Ok, I have already changed `ObjectContainer`.